gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java

```diff
@@ -0,0 +1,191 @@
+   *        will be used which looks for the {@link ConfigurationKeys#JOB_LAUNCHER_TYPE_KEY}
+   *        in the system configuration.
+   * @param jobExecStateListener  an optional listener to listen for state changes in the execution.
+   * @param log                   an optional logger to be used; if none is specified, a default one
+   *                              will be instantiated.
+   */
+  public JobLauncherExecutionDriver(Configurable sysConfig, JobSpec jobSpec,
+      Optional<JobLauncherFactory.JobLauncherType> jobLauncherType,
+      Optional<Logger> log) {
+    _log = log.isPresent() ? log.get() : LoggerFactory.getLogger(getClass());
+    _sysConfig = sysConfig;
+    _jobSpec = jobSpec;
+    _jobExec = JobExecutionUpdatable.createFromJobSpec(jobSpec);
+    _callbackDispatcher = new JobExecutionStateListeners(_log);
+    _jobState = new JobExecutionState(_jobSpec, _jobExec,
+                                      Optional.<JobExecutionStateListener>of(_callbackDispatcher));
+    _legacyLauncher =
+        createLauncher(jobLauncherType.isPresent() ?
+                      Optional.of(jobLauncherType.get().toString()) :
+                      Optional.<String>absent());
+  }
+
+  private JobLauncher createLauncher(Optional<String> jobLauncherType) {
+    if (jobLauncherType.isPresent()) {
+      return JobLauncherFactory.newJobLauncher(_sysConfig.getConfigAsProperties(),
+             _jobSpec.getConfigAsProperties(), jobLauncherType.get());
+    }
+    else {
+      _log.info("Creating auto jobLauncher for " + _jobSpec);
+      try {
+        return JobLauncherFactory.newJobLauncher(_sysConfig.getConfigAsProperties(),
+             _jobSpec.getConfigAsProperties());
+      } catch (Exception e) {
+        throw new RuntimeException("JobLauncher creation failed: " + e, e);
+      }
+    }
+  }
+
+  @Override
+  public JobExecution getJobExecution() {
+    return _jobExec;
+  }
+
+  @Override
+  public JobExecutionStatus getJobExecutionStatus() {
+    return _jobState;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    _log.info("Starting " + getClass().getSimpleName());
+    _legacyLauncher.launchJob(new MyJobListener());
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    _log.info("Shutting down " + getClass().getSimpleName());
+    if (null != _jobContext) {
+      switch (_jobContext.getJobState().getState()) {
+        case PENDING:
+        case RUNNING: {
+          // We have to pass another listener instance as launcher does not store the listener used
+          // in launchJob()
+          _legacyLauncher.cancelJob(new MyJobListener());
+          break;
+        }
+        case SUCCESSFUL:
+        case FAILED:
+        case COMMITTED:
+        case CANCELLED: {
+          // Nothing to do
+          break;
+        }
+      }
+    }
+
+    _legacyLauncher.close();
+  }
+
+  class MyJobListener extends AbstractJobListener {
```

Comment:
Can you change the name of the class? Maybe `JobListenerToJobStateBridge` or something more meaningful?